Add/Remove entities from systems when when its component bitmask changes

waldnercharles opened the pull request:
Figured I'd open this in case its helpful. I shamelessly stole the tests from https://github.com/empyreanx/pico_headers/pull/11. Feel free to close if you had something else in mind, or if there's something that's better for performance.

empyreanx commented:
For some reason the `tests.yml` workflow is crashing on the mingw32/mingw64 jobs. It is difficult to see which test is actually failing so I'm now printing the `pwd` in the job commands. Please commit the revised `tests.yml` so that we can deduce the problem.

Otherwise I have no qualms with your PR, it's a simple solution to the problem. It's possible that some optimizations can be made using some bitset magic. I will give it some thought, but I think this is a great place to start.

empyreanx commented:
When I have a moment I will try the mingw32/mingw64 tests in MSYS2.

waldnercharles commented:
I'm running msys2 ucrt. I'll try running mingw64 to see if it's my test.

empyreanx commented:
I tried to replicate the problem in MSYS2 without success. I'll try a few more things tomorrow. The fact that this is happening is very peculiar since the ECS tests for the other platforms run successfully, and all of the other MinGW tests run successfully across all platforms. If it comes down to it, I'll create a new project and do some brute force debugging.

empyreanx commented:
One thing to consider, it could be a problem with the tests.

waldnercharles commented:
As a sanity check, I'll try running main with no changes tonight, and then the tests without my new test changes.

empyreanx commented:
Valgrind found the following:

```
==62292== 1 errors in context 1 of 2:
==62292== Use of uninitialised value of size 8
==62292==    at 0x10B85B: ecs_sparse_set_find (pico_ecs.h:1329)
==62292==    by 0x10B8DB: ecs_sparse_set_remove (pico_ecs.h:1339)
==62292==    by 0x10AA9E: ecs_add (pico_ecs.h:927)
==62292==    by 0x10DC81: test_queue_remove_system (main.c:606)
==62292==    by 0x109418: pu_run_test (pico_unit.h:302)
==62292==    by 0x10E240: suite_ecs (main.c:732)
==62292==    by 0x1095C0: pu_run_suite (pico_unit.h:369)
==62292==    by 0x10E2B9: main (main.c:741)
```
Hopefully this is our culprit.

empyreanx commented:
Valgrind reveals this:

```
==62998== 1 errors in context 1 of 2:
==62998== Use of uninitialised value of size 8
==62998==    at 0x10B85B: ecs_sparse_set_find (pico_ecs.h:1329)
==62998==    by 0x10B8DB: ecs_sparse_set_remove (pico_ecs.h:1339)
==62998==    by 0x10AA9E: ecs_add (pico_ecs.h:927)
==62998==    by 0x10DC81: test_queue_remove_system (main.c:607)
==62998==    by 0x109418: pu_run_test (pico_unit.h:302)
==62998==    by 0x10E240: suite_ecs (main.c:733)
==62998==    by 0x1095C0: pu_run_suite (pico_unit.h:369)
==62998==    by 0x10E2B9: main (main.c:742)
==62998== 
==62998== 
==62998== 1 errors in context 2 of 2:
==62998== Conditional jump or move depends on uninitialised value(s)
==62998==    at 0x10B835: ecs_sparse_set_find (pico_ecs.h:1329)
==62998==    by 0x10B8DB: ecs_sparse_set_remove (pico_ecs.h:1339)
==62998==    by 0x10AA9E: ecs_add (pico_ecs.h:927)
==62998==    by 0x10DC81: test_queue_remove_system (main.c:607)
==62998==    by 0x109418: pu_run_test (pico_unit.h:302)
==62998==    by 0x10E240: suite_ecs (main.c:733)
==62998==    by 0x1095C0: pu_run_suite (pico_unit.h:369)
==62998==    by 0x10E2B9: main (main.c:742)
```

empyreanx commented:
Valgrind reveals this:

```
==63165== 1 errors in context 1 of 2:
==63165== Use of uninitialised value of size 8
==63165==    at 0x10B85B: ecs_sparse_set_find (pico_ecs.h:1329)
==63165==    by 0x10B8DB: ecs_sparse_set_remove (pico_ecs.h:1339)
==63165==    by 0x10AA9E: ecs_add (pico_ecs.h:927)
==63165==    by 0x10DC81: test_queue_remove_system (main.c:606)
==63165==    by 0x109418: pu_run_test (pico_unit.h:302)
==63165==    by 0x10E240: suite_ecs (main.c:732)
==63165==    by 0x1095C0: pu_run_suite (pico_unit.h:369)
==63165==    by 0x10E2B9: main (main.c:741)
==63165== 
==63165== 
==63165== 1 errors in context 2 of 2:
==63165== Conditional jump or move depends on uninitialised value(s)
==63165==    at 0x10B835: ecs_sparse_set_find (pico_ecs.h:1329)
==63165==    by 0x10B8DB: ecs_sparse_set_remove (pico_ecs.h:1339)
==63165==    by 0x10AA9E: ecs_add (pico_ecs.h:927)
==63165==    by 0x10DC81: test_queue_remove_system (main.c:606)
==63165==    by 0x109418: pu_run_test (pico_unit.h:302)
==63165==    by 0x10E240: suite_ecs (main.c:732)
==63165==    by 0x1095C0: pu_run_suite (pico_unit.h:369)
==63165==    by 0x10E2B9: main (main.c:741)
```

waldnercharles commented:
Just looking from my phone, but, that's likely my doing. I assumed ecs_remove was safe to call even if the id didn't exist in the sparse set.

From a cursory look, my guess is we probably need to add a bounds check on the sparse set to make sure id is less than capacity.

empyreanx commented:
`ecs_sparse_set_remove` calls `ecs_sparse_set_find`, so there is no need to do a bounds check.

empyreanx commented:
Actually, perhaps I'm could be wrong. Maybe there is a need to do a bounds check in `ecs_sparse_set_find`

waldnercharles commented:
No worries. I added the bounds check and it seems to have [completed successfully](https://github.com/waldnercharles/pico_headers/actions/runs/13277689286) in my fork. (Still using windows-latest for good measure, not windows-2025)

Thanks for running valgrind; I didn't even think of it.

empyreanx commented:
I was able to do a bit of optimization. Naturally, the changes to `ecs_add` and `ecs_remove` will incur _some_ performance hit. The three systems benchmark saw run time increase 22%. I've been been able to get it down to around 6% by checking if the system has any exclusions. I'll push a commit later on this evening.

empyreanx commented:
I think everything is about ready to go. I would, however, appreciate any comments you have regarding the optimization. It was designed to make the common case fast (i.e. no exclusions). If you agree with the change, I'll merge it right away.

waldnercharles commented:
Looks great! All of the negative logic was hurting my brain, but, I think it makes sense. ie. When removing a component, we only need to add an entity to a system if that system could have been excluding the component we just removed. 

We might get an optimization for the exclude-case by doing something similar: maybe a bitset test on the exclude bits for the component we just removed/added? I'd need to benchmark.

empyreanx commented:
It might be helpful to think of the `ecs_remove` case in these terms: if a system has no exclusions, then there is no need to perform any dynamic operations at all, and thus, no need to add the entity to the system.
